File: app/routers/categories.py
"""
FastAPI TeuxDeux Clone - Categories Router
CRUD operations for todo categories
"""

from fastapi import APIRouter, Depends, HTTPException, Path
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.dependencies import get_db_session
from app.models import (
    Category, Todo, CreateCategoryRequest, UpdateCategoryRequest, 
    CategoryResponse, APIResponse
)

router = APIRouter()

@router.get("/categories", response_model=APIResponse)
async def get_categories(
    db: AsyncSession = Depends(get_db_session)
):
    """Get all categories"""
    
    query = select(Category).order_by(Category.sort_order.asc(), Category.name.asc())
    result = await db.execute(query)
    categories = result.scalars().all()
    
    category_responses = [CategoryResponse.from_orm(cat) for cat in categories]
    
    return APIResponse(
        success=True,
        data={"categories": [cat.dict() for cat in category_responses]}
    )

@router.post("/categories", response_model=APIResponse)
async def create_category(
    category_data: CreateCategoryRequest,
    db: AsyncSession = Depends(get_db_session)
):
    """Create a new category"""
    
    # Set default color if not provided
    color = category_data.color if category_data.color else "#6b7280"
    
    category = Category(
        name=category_data.name,
        color=color
    )
    
    db.add(category)
    await db.commit()
    await db.refresh(category)
    
    return APIResponse(
        success=True,
        message="Category created successfully",
        data={"id": category.id}
    )

@router.put("/categories/{category_id}", response_model=APIResponse)
async def update_category(
    category_id: int = Path(..., description="Category ID"),
    category_data: UpdateCategoryRequest = None,
    db: AsyncSession = Depends(get_db_session)
):
    """Update an existing category"""
    
    # Get the category
    category = await db.get(Category, category_id)
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    
    # Update fields that are provided
    update_data = category_data.dict(exclude_unset=True)
    if not update_data:
        raise HTTPException(status_code=400, detail="No fields to update")
    
    for field, value in update_data.items():
        setattr(category, field, value)
    
    await db.commit()
    
    return APIResponse(
        success=True,
        message="Category updated successfully"
    )

@router.delete("/categories/{category_id}", response_model=APIResponse)
async def delete_category(
    category_id: int = Path(..., description="Category ID"),
    db: AsyncSession = Depends(get_db_session)
):
    """Delete a category"""
    
    category = await db.get(Category, category_id)
    if not category:
        raise HTTPException(status_code=404, detail="Category not found")
    
    # Check if category is in use
    count_query = select(func.count(Todo.id)).where(Todo.category_id == category_id)
    result = await db.execute(count_query)
    count = result.scalar()
    
    if count > 0:
        raise HTTPException(
            status_code=400, 
            detail=f"Cannot delete category: {count} todos are using this category"
        )
    
    await db.delete(category)
    await db.commit()
    
    return APIResponse(
        success=True,
        message="Category deleted successfully"
    )

File: go-app/internal/handlers/categories.go
package handlers

import (
	"fmt"
	"net/http"
	"strconv"

	"teuxdeux-clone/internal/models"

	"github.com/gin-gonic/gin"
)

// GetCategories returns all categories
func (h *Handler) GetCategories(c *gin.Context) {
	categories, err := h.getCategories()
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Error:   "Failed to get categories: " + err.Error(),
		})
		return
	}
	
	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Data:    categories,
	})
}

// CreateCategory creates a new category
func (h *Handler) CreateCategory(c *gin.Context) {
	var req models.CreateCategoryRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Error:   "Invalid request body: " + err.Error(),
		})
		return
	}
	
	// Set default color if not provided
	if req.Color == "" {
		req.Color = "#6b7280" // Default gray color
	}
	
	query := `
		INSERT INTO categories (name, color)
		VALUES (?, ?)
	`
	
	result, err := h.db.Exec(query, req.Name, req.Color)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Error:   "Failed to create category: " + err.Error(),
		})
		return
	}
	
	categoryID, _ := result.LastInsertId()
	
	c.JSON(http.StatusCreated, models.APIResponse{
		Success: true,
		Message: "Category created successfully",
		Data:    map[string]interface{}{"id": categoryID},
	})
}

// UpdateCategory updates an existing category
func (h *Handler) UpdateCategory(c *gin.Context) {
	categoryID, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Error:   "Invalid category ID",
		})
		return
	}
	
	var req models.UpdateCategoryRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Error:   "Invalid request body: " + err.Error(),
		})
		return
	}
	
	// Build dynamic update query
	setParts := []string{}
	args := []interface{}{}
	
	if req.Name != nil {
		setParts = append(setParts, "name = ?")
		args = append(args, *req.Name)
	}
	if req.Color != nil {
		setParts = append(setParts, "color = ?")
		args = append(args, *req.Color)
	}
	if req.SortOrder != nil {
		setParts = append(setParts, "sort_order = ?")
		args = append(args, *req.SortOrder)
	}
	
	if len(setParts) == 0 {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Error:   "No fields to update",
		})
		return
	}
	
	query := "UPDATE categories SET " + setParts[0]
	for i := 1; i < len(setParts); i++ {
		query += ", " + setParts[i]
	}
	query += " WHERE id = ?"
	args = append(args, categoryID)
	
	_, err = h.db.Exec(query, args...)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Error:   "Failed to update category: " + err.Error(),
		})
		return
	}
	
	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Category updated successfully",
	})
}

// DeleteCategory deletes a category
func (h *Handler) DeleteCategory(c *gin.Context) {
	categoryID, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Error:   "Invalid category ID",
		})
		return
	}
	
	// Check if category is in use
	var count int
	err = h.db.QueryRow("SELECT COUNT(*) FROM todos WHERE category_id = ?", categoryID).Scan(&count)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Error:   "Failed to check category usage: " + err.Error(),
		})
		return
	}
	
	if count > 0 {
		c.JSON(http.StatusBadRequest, models.APIResponse{
			Success: false,
			Error:   fmt.Sprintf("Cannot delete category: %d todos are using this category", count),
		})
		return
	}
	
	_, err = h.db.Exec("DELETE FROM categories WHERE id = ?", categoryID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.APIResponse{
			Success: false,
			Error:   "Failed to delete category: " + err.Error(),
		})
		return
	}
	
	c.JSON(http.StatusOK, models.APIResponse{
		Success: true,
		Message: "Category deleted successfully",
	})
}